content/posts/2018-01-10-kotlin-null.md

```diff
@@ -0,0 +1,249 @@
+    fun calculate(argument: Argument)
+
+    class CorporateImpl : Calculator {
+        override fun calculate(argument: Argument) {
+            CorporateCalculator().calculate(argument)
+        }
+    }
+}
+```
+
+As you see, we hide the `CorporateCalculator` behind a Kotlin-based abstraction
+which will save us from passing `null` to the `calculate` method.
+
+# Making a Swift Turn
+
+The good news is — we are not alone. Like, on the planet!
+
+The Java and Kotlin situation is kind of similar to the Objective-C and Swift one.
+Objective-C objects are essentially C pointers, which can be `NULL` or `nil` in Objective-C terms.
+In other words, these objects are implicitly nullable.
+On another hand, Swift requires all values to have nullability defined explicitly.
+
+Sounds exactly like our case, right?
+
+Apple provides Nullability Annotations for Objective-C as well — `_Nullable` + `nullable` and `_Nonnull` + `nonnull`.
+You use them — Swift fetches them using the Objective-C and Swift bridging.
+But what happens when you don’t use annotations in your Objective-C code? This is where things are getting interesting.
+
+Instead of taking the Kotlin-like path where all Java values without annotations on them
+are getting almost a blind check (pass `null` values freely as arguments and get results as non-nullable),
+Swift declares them as [_implicitly unwrapped optionals_](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/TheBasics.html).
+In other words, Swift kind of unwraps values for you,
+but you are doing it at your own risk since it is basically an equivalent of doing `nullable!!` in Kotlin.
+At the same time, it gives you a benefit of the non-warning-hostile environment.
+Just imagine if _all_ Java arguments and results would be nullable! It would be honest, but kind of gross to use.
+Unfortunately, it is not possible to define such types using Kotlin at this point.
+
+## Side Road: Swift Optionals
+
+Turns out Swift does not have `null` pointers. At all!
+
+```swift
+let text: String? = nil
+```
+
+`text` there is not actually `null` (`nil`). Let’s change it a bit to show its true nature.
+
+```swift
+let text: Optional<String> = Optional.none
+
+enum Optional<Wrapped> {
+    case none
+    case some(Wrapped)
+}
+```
+
+Pretty cool, right? The Swift compiler makes all nullable values `Optional` under the hood.
+In other words, `?` and `nil` are just syntax sugar literals for
+[`enum Optional`](https://developer.apple.com/documentation/swift/optional),
+which is extremely similar to a Kotlin `sealed class Optional`
+like [Koptional](https://github.com/gojuno/koptional).
+Swift is open source so you can take a look at
+[the `Optional` implementation](https://github.com/apple/swift/blob/master/stdlib/public/core/Optional.swift)
+yourself since I’ve simplified it drastacially at the code sample above.
+
+Unfortunately, it is not really possible to change Kotlin behave the same way.
```

Comment:
I think you're missing the point here. The fact that nil and null are literals to Optional.none in Swift doesn't work in reverse direction for Objective-C, ie as far as I understand Optional.none won't be nill for Objective-C, or I'm wrong?

I guess the real point is that Kotlin doesn't have Optional in stdlib while Swift does

Reply:
I’ll double-check, but for me `Optional.none` will be `nil` in Objective-C.